File: src/igda/matchers/naive.py
from __future__ import annotations

from igda.core import AlgorithmInfo, MatchKind, MatchResult
from igda.matchers import utils


def _one_pattern(text: str, pat: str) -> tuple[list[dict[str, int]], int]:
    n, m = len(text), len(pat)
    if m == 0 or n < m:
        return [], 0
    out: list[dict[str, int]] = []
    comparisons = 0
    for i in range(0, n - m + 1):
        j = 0
        while j < m:
            comparisons += 1
            if text[i + j] != pat[j]:
                break
            j += 1
        if j == m:
            out.append({"start": i, "end": i + m, "pattern": pat})
    return out, comparisons


class NaiveMatcher:
    info = AlgorithmInfo(
        id="naive",
        name="Naive / brute-force",
        match_kind=MatchKind.EXACT,
        time_complexity="O(n·m) per pattern; multiple patterns: O(Σ|Pi|·n)",
        space_complexity="O(1) auxiliary (excluding output list)",
        notes="Trivial; baseline. Retries from every shift; no preprocessing reuse.",
        supports_multi_pattern=True,
    )

    def match(self, text: str, patterns: list[str], **kwargs: object) -> MatchResult:
        text, patterns = utils.sanitize_text_patterns(text, patterns, allow_empty_pattern=False)
        all_rows: list[dict[str, int | str]] = []
        total_c = 0
        for pat in patterns:
            rows, c = _one_pattern(text, pat)
            all_rows.extend(rows)  # type: ignore[assignment]
            total_c += c
        return MatchResult(
            algorithm_id=self.info.id,
            matches=all_rows,
            match_count=len(all_rows),
            extra={"character_comparisons": total_c},
        )

File: source/matchers/naive.c
/*
 * Naive / Brute-force string matching
 *
 * Time complexity : O(n * m) per pattern
 * Space complexity: O(1) auxiliary (excluding output)
 *
 * Trivial baseline: retries from every shift; no preprocessing reuse.
 *
 * IGDA reference: src/igda/matchers/naive.py
 */

#include <stdio.h>
#include <string.h>

typedef struct {
    int start;
    int end;
} Match;

/*
 * Find all exact occurrences of pattern in text.
 * Writes match start indices into starts[] (at most max_matches entries).
 * Returns number of matches; *comparisons receives character comparisons.
 */
int naive_match(const char *text, const char *pattern,
                int *starts, int max_matches, int *comparisons) {
    int n = (int)strlen(text);
    int m = (int)strlen(pattern);
    int found = 0;
    *comparisons = 0;

    if (m == 0 || n < m)
        return 0;

    for (int i = 0; i <= n - m; i++) {
        int j = 0;
        while (j < m) {
            (*comparisons)++;
            if (text[i + j] != pattern[j])
                break;
            j++;
        }
        if (j == m && found < max_matches) {
            starts[found++] = i;
        }
    }
    return found;
}

#ifdef NAIVE_DEMO
int main(void) {
    const char *text = "ACGTACGTACGT";
    const char *pat = "ACGT";
    int starts[32];
    int cmp = 0;
    int k = naive_match(text, pat, starts, 32, &cmp);

    printf("Naive match: text=\"%s\" pattern=\"%s\"\n", text, pat);
    printf("Matches: %d, comparisons: %d\n", k, cmp);
    for (int i = 0; i < k; i++)
        printf("  start=%d end=%d\n", starts[i], starts[i] + (int)strlen(pat));
    return 0;
}
#endif
